// @builder.io/angular/esm2022/app/modules/builder/components/builder-content/builder-content.component.mjs
import { Component, EventEmitter, Output, Input, ChangeDetectionStrategy, } from '@angular/core';
import { Builder } from '@builder.io/sdk';
import { BuilderService } from '../../services/builder.service';
import * as i0 from "@angular/core";
import * as i1 from "../builder-component/builder-component.service";
export class BuilderContentComponent {
    element;
    builderComponentService;
    constructor(element, builderComponentService) {
        this.element = element;
        this.builderComponentService = builderComponentService;
        builderComponentService.contentComponentInstance = this;
    }
    modelName;
    useHtml = false;
    data = {};
    hydrate = true;
    prerender = true;
    set content(content) {
        const currentContent = this._content;
        this._content = content;
        const { contentDirectiveInstance } = this.builderComponentService;
        if (!currentContent && content && contentDirectiveInstance) {
            if (!contentDirectiveInstance.requesting) {
                contentDirectiveInstance.reset();
            }
        }
    }
    get content() {
        return this._content;
    }
    _content;
    findAndRunScripts() {
        if (!Builder.isBrowser) {
            return;
        }
        const el = this.element.nativeElement;
        if (el) {
            const scripts = el.getElementsByTagName('script');
            for (let i = 0; i < scripts.length; i++) {
                const script = scripts[i];
                if (script.src) {
                    const newScript = document.createElement('script');
                    newScript.async = true;
                    newScript.src = script.src;
                    document.head.appendChild(newScript);
                }
                else {
                    try {
                        new Function(script.innerText)();
                    }
                    catch (error) {
                        console.warn('Builder custom code component error:', error);
                    }
                }
            }
        }
    }
    options = null;
    contentLoad = new EventEmitter();
    contentError = new EventEmitter();
    get editingMode() {
        return Builder.editingPage;
    }
    ngOnInit() {
        const modelName = this.element.nativeElement &&
            this.element.nativeElement.getAttribute &&
            this.element.nativeElement.getAttribute('builder-model');
        if (modelName) {
            // FIXME: doesn't work on server!
            this.modelName = modelName;
            BuilderService.componentInstances[modelName] = this;
        }
    }
    ngOnDestroy() {
        if (this.modelName) {
            if (BuilderService.componentInstances[this.modelName] === this) {
                delete BuilderService.componentInstances[this.modelName];
            }
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: BuilderContentComponent, deps: [{ token: i0.ElementRef }, { token: i1.BuilderComponentService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.0.6", type: BuilderContentComponent, selector: "builder-content", inputs: { useHtml: "useHtml", data: "data", hydrate: "hydrate", prerender: "prerender", content: "content", options: "options" }, outputs: { contentLoad: "contentLoad", contentError: "contentError" }, ngImport: i0, template: "<ng-content> </ng-content>\n", styles: [":host{display:block}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: BuilderContentComponent, decorators: [{
            type: Component,
            args: [{ selector: 'builder-content', changeDetection: ChangeDetectionStrategy.OnPush, template: "<ng-content> </ng-content>\n", styles: [":host{display:block}\n"] }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i1.BuilderComponentService }], propDecorators: { useHtml: [{
                type: Input
            }], data: [{
                type: Input
            }], hydrate: [{
                type: Input
            }], prerender: [{
                type: Input
            }], content: [{
                type: Input
            }], options: [{
                type: Input
            }], contentLoad: [{
                type: Output
            }], contentError: [{
                type: Output
            }] } });
