// @builder.io/angular/esm2022/app/modules/builder/builder.module.mjs
import './polyfills/custom-elements-es5-adapter';
import { Builder } from '@builder.io/sdk';
import { NgModule, Inject, PLATFORM_ID } from '@angular/core';
import { CommonModule, isPlatformBrowser } from '@angular/common';
import { BuilderContentComponent } from './components/builder-content/builder-content.component';
import { BuilderContentDirective } from './directives/builder-content.directive';
import { BUILDER_API_KEY, BuilderService } from './services/builder.service';
import { BuilderContentService } from './services/builder-content.service';
import { BuilderBlocksComponent } from './components/builder-blocks/builder-blocks.component';
import { BuilderBlocksOutletComponent } from './components/builder-blocks-outlet/builder-blocks-outlet.component';
import { BuilderComponentComponent } from './components/builder-component/builder-component.component';
import { BuilderComponentService } from './components/builder-component/builder-component.service';
import * as i0 from "@angular/core";
Builder.isStatic = true;
Builder.sdkInfo = {
    name: 'angular',
    version: '5.0.2',
};
if (typeof window !== 'undefined') {
    window.parent?.postMessage({
        type: 'builder.isAngularGen1Sdk',
        data: {
            // @ts-ignore
            version: '5.0.2',
        },
    }, '*');
}
const directives = [BuilderContentDirective];
const components = [
    BuilderContentComponent,
    BuilderBlocksComponent,
    BuilderComponentComponent,
    BuilderBlocksOutletComponent,
];
export class BuilderModule {
    platformId;
    constructor(injector, platformId) {
        this.platformId = platformId;
        if (isPlatformBrowser(platformId)) {
            // This cannot use a normal import, via https://github.com/angular/angular/issues/24551
            // Cannot use require, so use import. This otherwise breaks at runtime
            import('@angular/elements').then(({ createCustomElement }) => {
                for (const component of Builder.components) {
                    if (component.class &&
                        component.type === 'angular' &&
                        component.tag &&
                        typeof customElements.get(component.tag) === 'undefined') {
                        try {
                            const Element = createCustomElement(component.class, { injector });
                            // Register the custom element with the browser.
                            customElements.define(component.tag, Element);
                        }
                        catch (err) {
                            console.warn('Could not make angular element:', component.class);
                        }
                    }
                }
            });
        }
    }
    static forRoot(apiKey) {
        return {
            ngModule: BuilderModule,
            providers: [
                {
                    provide: BUILDER_API_KEY,
                    useValue: apiKey,
                },
            ],
        };
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: BuilderModule, deps: [{ token: i0.Injector }, { token: PLATFORM_ID }], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.6", ngImport: i0, type: BuilderModule, declarations: [BuilderContentComponent,
            BuilderBlocksComponent,
            BuilderComponentComponent,
            BuilderBlocksOutletComponent, BuilderContentDirective], imports: [CommonModule], exports: [BuilderContentComponent,
            BuilderBlocksComponent,
            BuilderComponentComponent,
            BuilderBlocksOutletComponent, BuilderContentDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: BuilderModule, providers: [BuilderService, BuilderContentService, BuilderComponentService], imports: [CommonModule] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: BuilderModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    providers: [BuilderService, BuilderContentService, BuilderComponentService],
                    declarations: [components, directives],
                    exports: [components, directives],
                }]
        }], ctorParameters: () => [{ type: i0.Injector }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [PLATFORM_ID]
                }] }] });
