// @builder.io/angular/esm2022/app/modules/builder/services/builder.service.mjs
import { Injectable, Inject, Optional } from '@angular/core';
import { RESPONSE, REQUEST } from '../../../../express.tokens';
import { Builder } from '@builder.io/sdk';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export const BUILDER_API_KEY = 'BUILDER_API_KEY'; // new InjectionToken<string>('BUILDER_API_KEY');
export const EXPRESS_REQUEST = 'EXPRESS_REQUEST'; // new InjectionToken<Request>('EXPRESS_REQUEST');
export const EXPRESS_RESPONSE = 'EXPRESS_RESPONSE'; // new InjectionToken<Response>('EXPRESS_RESPONSE');
export class BuilderService extends Builder {
    expressRequest;
    expressEngineRequest;
    expressResponse;
    expressEngineResponse;
    http;
    static componentInstances = {};
    autoTrack = !this.isDevelopment;
    // TODO: set this for QA
    get isDevelopment() {
        // Automatic determining of development environment
        return (Builder.isIframe ||
            (Builder.isBrowser && (location.hostname === 'localhost' || location.port !== '')));
    }
    constructor(apiKey, expressRequest, expressEngineRequest, expressResponse, expressEngineResponse, http) {
        super(apiKey, expressEngineRequest || expressRequest, expressEngineResponse || expressResponse);
        this.expressRequest = expressRequest;
        this.expressEngineRequest = expressEngineRequest;
        this.expressResponse = expressResponse;
        this.expressEngineResponse = expressEngineResponse;
        this.http = http;
        if (this.expressEngineRequest) {
            this.expressRequest = this.expressEngineRequest;
        }
        if (this.expressRequest) {
            this.request = this.expressRequest;
        }
        if (this.expressEngineResponse) {
            this.expressResponse = this.expressEngineResponse;
        }
        if (this.expressResponse) {
            this.response = this.expressResponse;
        }
        if (apiKey) {
            this.apiVersion = 'v3';
            this.init(apiKey);
        }
        if (!Builder.isBrowser && !this.request) {
            console.warn('No express request set! Builder cannot target appropriately without this, ' +
                'please contact support@builder.io to learn how to set this as required');
        }
    }
    // (override)
    requestUrl(url) {
        if (this.http) {
            return this.http.get(url).toPromise();
        }
        else {
            return super.requestUrl(url);
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: BuilderService, deps: [{ token: BUILDER_API_KEY, optional: true }, { token: EXPRESS_REQUEST, optional: true }, { token: REQUEST, optional: true }, { token: EXPRESS_RESPONSE, optional: true }, { token: RESPONSE, optional: true }, { token: i1.HttpClient, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: BuilderService });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: BuilderService, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [BUILDER_API_KEY]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [EXPRESS_REQUEST]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [REQUEST]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [EXPRESS_RESPONSE]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [RESPONSE]
                }] }, { type: i1.HttpClient, decorators: [{
                    type: Optional
                }] }] });
