// @builder.io/angular/esm2022/app/modules/builder/components/builder-blocks-outlet/builder-blocks-outlet.component.mjs
import { Component, Input, ChangeDetectionStrategy, } from '@angular/core';
import { Builder } from '@builder.io/sdk';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
export class BuilderBlocksOutletComponent {
    domSanitizer;
    builderBlock;
    builderState;
    blocks;
    renderOnChange = true;
    dataPath;
    lastInnerHtml = '';
    lastInnerHtmlSanitized;
    get options() {
        return {
            child: true,
            parentElementId: this.builderBlock.id,
            blocks: this.blocks,
            dataPath: this.dataPath,
        };
    }
    get key() {
        return this.builderBlock.id + this.dataPath;
    }
    get innerHtml() {
        const html = this._innerHtml;
        if (html === this.lastInnerHtml) {
            return this.lastInnerHtmlSanitized || '';
        }
        this.lastInnerHtml = html;
        this.lastInnerHtmlSanitized = this.domSanitizer.bypassSecurityTrustHtml(html);
        return this.lastInnerHtmlSanitized;
    }
    get _innerHtml() {
        return `<builder-blocks-slot key="${this.key}"></builder-blocks-slot>`;
    }
    constructor(domSanitizer) {
        this.domSanitizer = domSanitizer;
    }
    ngAfterViewInit() {
        if (Builder.isBrowser) {
            this.triggerstateChange();
        }
    }
    async triggerstateChange() {
        const query = `builder-blocks-slot[key="${this.key}"]`;
        const element = document.querySelector(query);
        if (element) {
            await customElements.whenDefined('builder-blocks-slot');
            element.setProps(this.options, this.builderState);
        }
    }
    ngOnChanges(changes) {
        if (changes.blocks && this.renderOnChange) {
            this.triggerstateChange();
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: BuilderBlocksOutletComponent, deps: [{ token: i1.DomSanitizer }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.0.6", type: BuilderBlocksOutletComponent, selector: "builder-blocks-outlet", inputs: { builderBlock: "builderBlock", builderState: "builderState", blocks: "blocks", renderOnChange: "renderOnChange", dataPath: "dataPath" }, usesOnChanges: true, ngImport: i0, template: "<div [innerHTML]=\"innerHtml\"></div>\n", styles: [":host{display:block}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: BuilderBlocksOutletComponent, decorators: [{
            type: Component,
            args: [{ selector: 'builder-blocks-outlet', changeDetection: ChangeDetectionStrategy.OnPush, template: "<div [innerHTML]=\"innerHtml\"></div>\n", styles: [":host{display:block}\n"] }]
        }], ctorParameters: () => [{ type: i1.DomSanitizer }], propDecorators: { builderBlock: [{
                type: Input
            }], builderState: [{
                type: Input
            }], blocks: [{
                type: Input
            }], renderOnChange: [{
                type: Input
            }], dataPath: [{
                type: Input
            }] } });
